### PR TITLE
Refuse merge until all required status checks success

### LINKED ISSUE #3588: Disallow merging until the CI bot reports about success build
This is supported for the github and travis CI, for example. So, using a github and travis CI you may configure it so that it is impossible to merge the pull requests before the CI bot allows you.

### PR DESCRIPTION
**Pull request by lunny**
Will fix #3588 and #996

 **Screenshot 1** protected branch settings:

image

**Screenshot 2** all required status checks pass:

image

**Screenshot 3** some required status checks failed and user is not repository administrator:

image

**Screenshot 4** some required status checks failed but user is the repository administrator, it still can be merged:

image

### CODE REVIEW

**Comment by techknowlogick** [comment]
Can admin bypass checks?

This is what I see as an example on github:
Screen Shot 2019-07-15 at 11 22 22 PM

**Comment by lunny** [comment]
@techknowlogick Will try to add admin bypass checks .

**Comment by techknowlogick** [comment]
Admin bypass checks could be for different PR

**Comment by lunny** [comment]
@techknowlogick That's OK for quick review. I'll send another PR after this merged.

File: models/branches.go [hunk 1]
```diff
@@ -35,6 +35,8 @@ type ProtectedBranch struct {
 	EnableMergeWhitelist      bool           `xorm:"NOT NULL DEFAULT false"`
 	MergeWhitelistUserIDs     []int64        `xorm:"JSON TEXT"`
 	MergeWhitelistTeamIDs     []int64        `xorm:"JSON TEXT"`
+	EnableStatusCheck         bool           `xorm:"NOT NULL DEFAULT false"`
```
Comment (techknowlogick) [inline models/branches.go]: Can you make a migration to add these two columns?

Comment (lunny) [reply models/branches.go]: done.

File: options/locale/locale_en-US.ini [hunk 2]
```diff
@@ -1294,6 +1296,8 @@ settings.protect_merge_whitelist_committers = Enable Merge Whitelist
 settings.protect_merge_whitelist_committers_desc = Allow only whitelisted users or teams to merge pull requests into this branch.
 settings.protect_merge_whitelist_users = Whitelisted users for merging:
 settings.protect_merge_whitelist_teams = Whitelisted teams for merging:
+settings.protect_check_status_contexts = Status checks found in the last week for this repository
+settings.protect_check_status_contexts_desc = Require status checks to pass before merging Choose which status checks must pass before branches can be merged into a branch that matches this rule. When enabled, commits must first be pushed to another branch, then merged or pushed directly to a branch that matches this rule after status checks have passed.
```
Comment (kolaente) [inline options/locale/locale_en-US.ini:1318]:
```suggestion
settings.protect_check_status_contexts_desc = Choose which status checks must pass before a pull-request can be merged.
```

File: templates/repo/settings/protected_branch.tmpl [hunk 3]
```diff
@@ -104,6 +104,38 @@
 					{{end}}
 					</div>
 
+					<div class="field">
+						<div class="ui checkbox">
+							<input class="enable-statuscheck" name="enable_status_check" type="checkbox" data-target="#statuscheck_contexts_box" {{if .Branch.EnableStatusCheck}}checked{{end}}>
+							<label>{{.i18n.Tr "repo.settings.protect_check_status_contexts"}}</label>
```
Comment (kolaente) [inline templates/repo/settings/protected_branch.tmpl:110]: Since this is a headline, it should be something like `Require passing checks before merging` and not `Status checks found in the last week for this repository`

File: routers/repo/pull.go [hunk 4]
```diff
@@ -582,6 +600,44 @@ func MergePullRequest(ctx *context.Context, form auth.MergePullRequestForm) {
 		return
 	}
 
+	// check if all required status checks are successful
```
Comment (kolaente) [inline routers/repo/pull.go]: Since you're doing this check (the following 40 lines) the second time here (first time was in `routers/api/v1/repo/pull.go:581`), could you put that in a seperate function?

File: routers/repo/setting_protected_branch.go [hunk 5]
```diff
@@ -125,6 +126,15 @@ func SettingsProtectedBranch(c *context.Context) {
 	c.Data["whitelist_users"] = strings.Join(base.Int64sToStrings(protectBranch.WhitelistUserIDs), ",")
 	c.Data["merge_whitelist_users"] = strings.Join(base.Int64sToStrings(protectBranch.MergeWhitelistUserIDs), ",")
 	c.Data["approvals_whitelist_users"] = strings.Join(base.Int64sToStrings(protectBranch.ApprovalsWhitelistUserIDs), ",")
+	c.Data["branch_status_check_contexts"], _ = models.FindRepoStatusCheckContexts(c.Repo.Repository.ID, 4*7*24*time.Hour) // Find last week status check contexts
```
Comment (kolaente) [inline routers/repo/setting_protected_branch.go]: Isn't `4*7*24*time.Hour` a month and not a week?

File: routers/repo/setting_protected_branch.go [hunk 6]
```diff
@@ -186,6 +196,13 @@ func SettingsProtectedBranchPost(ctx *context.Context, f auth.ProtectBranchForm)
 		if strings.TrimSpace(f.MergeWhitelistTeams) != "" {
 			mergeWhitelistTeams, _ = base.StringsToInt64s(strings.Split(f.MergeWhitelistTeams, ","))
 		}
+		if strings.TrimSpace(f.MergeWhitelistTeams) != "" {
```
Comment (kolaente) [inline routers/repo/setting_protected_branch.go]: Isn't this the same thing as the `if` just above?

**Review by kolaente — CHANGES_REQUESTED** [review]
What do you think about not using the statuses from the last week, but the last 10 or so in the settings dialog? If I won't push to a repo for some time, I wont see the checks I want to require.

**Comment by MarkusAmshove** [comment]
I have a question about the used identifier/context of the commit status.

Currently the Gitea Jenkins Plugins annotates a commit with a status in each job it is executing.

Since we're staging our PRs twice (once via PR into QA and once via PR into production), our `commit_status` table for a single commit has 3 entries:

- Context: REPONAME/ticketbranch
- Context: REPONAME/PR-(PR# to QA)
- Context: REPONAME/PR-(PR# to master)

1. Will this pollute the suggested checks in the branch protection settings?
2. Does the Jenkins plugin need to change to e.g. not include the PR number?
3. If a commit is within 2 PR's, would one PR build overwrite the status of the other PR?

**Comment by eduardosada** [comment]
Would be nice to have this :)

File: routers/repo/setting_protected_branch.go [hunk 7]
```diff
@@ -183,9 +207,10 @@ func SettingsProtectedBranchPost(ctx *context.Context, f auth.ProtectBranchForm)
 		if strings.TrimSpace(f.MergeWhitelistUsers) != "" {
 			mergeWhitelistUsers, _ = base.StringsToInt64s(strings.Split(f.MergeWhitelistUsers, ","))
 		}
-		if strings.TrimSpace(f.MergeWhitelistTeams) != "" {
-			mergeWhitelistTeams, _ = base.StringsToInt64s(strings.Split(f.MergeWhitelistTeams, ","))
-		}
```
Comment (guillep2k) [inline routers/repo/setting_protected_branch.go:188]: Why was this left out?

File: templates/repo/pulls/status.tmpl [hunk 8]
```diff
@@ -15,7 +15,12 @@
...
-            <div class="ui right">{{if .TargetURL}}<a href="{{.TargetURL}}">Details</a>{{end}}</div>
+            <div class="ui right">
+                {{if $.is_context_required}}
+                    {{if (call $.is_context_required .Context)}}<div class="ui label">Required</div>{{end}} 
+                {{end}}
+                <span class="ui">{{if .TargetURL}}<a href="{{.TargetURL}}">Details</a>{{end}}</span>
+            </div>
         </div>
     {{end}}
 {{end}}
```
Comment (guillep2k) [inline templates/repo/pulls/status.tmpl]: Newline is required here?

**Review by guillep2k — APPROVED** [review]
This was a though one to review! :grin:

File: routers/repo/setting_protected_branch.go (hunk 7 above)
Comment (lunny) [reply routers/repo/setting_protected_branch.go:188]: done.

File: templates/repo/pulls/status.tmpl (hunk 8 above)
Comment (lunny) [reply templates/repo/pulls/status.tmpl]: done.

File: options/locale/locale_en-US.ini (hunk 2 above)
Comment (lunny) [reply options/locale/locale_en-US.ini:1318]: The description comes from github originally.

File: templates/repo/settings/protected_branch.tmpl (hunk 3 above)
Comment (lunny) [reply templates/repo/settings/protected_branch.tmpl:110]: As above.

**Comment by lunny** [comment]
@kolaente All the description comes from github originally.

File: options/locale/locale_en-US.ini [hunk 9]
```diff
@@ -978,13 +978,15 @@ pulls.cannot_merge_work_in_progress = This pull request is marked as a work in p
...
 pulls.is_checking = "Merge conflict checking is in progress. Try again in few moments."
+pulls.required_status_check_failed = Some required checks were not successful.
 pulls.blocked_by_approvals = "This Pull Request doesn't have enough approvals yet. %d of %d approvals granted."
 pulls.can_auto_merge_desc = This pull request can be merged automatically.
 pulls.cannot_auto_merge_desc = This pull request cannot be merged automatically due to conflicts.
 pulls.cannot_auto_merge_helper = Merge manually to resolve the conflicts.
 pulls.no_merge_desc = This pull request cannot be merged because all repository merge options are disabled.
 pulls.no_merge_helper = Enable merge options in the repository settings or merge the pull request manually.
 pulls.no_merge_wip = This pull request can not be merged because it is marked as being a work in progress.
+pulls.no_merge_status_check = This pull request cannot be merged because not all required status checkes are successful.
```
Comment (guillep2k) [inline options/locale/locale_en-US.ini:993]:
```suggestion
pulls.no_merge_status_check = This pull request cannot be merged because some of the required status checks were unsuccessful.
```

File: options/locale/locale_en-US.ini [hunk 10]
```diff
@@ -1306,6 +1308,9 @@ settings.protect_merge_whitelist_committers = Enable Merge Whitelist
 settings.protect_merge_whitelist_committers_desc = Allow only whitelisted users or teams to merge pull requests into this branch.
 settings.protect_merge_whitelist_users = Whitelisted users for merging:
 settings.protect_merge_whitelist_teams = Whitelisted teams for merging:
+settings.protect_check_status_contexts = Enable Status Check
+settings.protect_check_status_contexts_desc = Require status checks to pass before merging Choose which status checks must pass before branches can be merged into a branch that matches this rule. When enabled, commits must first be pushed to another branch, then merged or pushed directly to a branch that matches this rule after status checks have passed.
```
Comment (guillep2k) [inline options/locale/locale_en-US.ini:1318]:
```suggestion
settings.protect_check_status_contexts_desc = Require status checks to pass before merging. Choose which status checks must pass before branches can be merged into a branch that matches this rule. When enabled, commits must first be pushed to another branch, then merged or pushed directly to a branch that matches this rule after status checks have passed.
```

**Review by guillep2k — CHANGES_REQUESTED** [review]

File: models/commit_status.go [hunk 11]
```diff
@@ -205,6 +206,26 @@ func GetLatestCommitStatus(repo *Repository, sha string, page int) ([]*CommitSta
...
 
+// FindRepoStatusCheckContexts returns repo's status check contexts
+func FindRepoStatusCheckContexts(repoID int64, before time.Duration) ([]string, error) {
+	start := timeutil.TimeStampNow().AddDuration(-before)
+	ids := make([]int64, 0, 10)
+	err := x.Table("commit_status").
+		Where("repo_id = ?", repoID).
+		And("updated_unix >= ?", start).
+		Select("max( id ) as id").
+		GroupBy("context_hash").OrderBy("max( id ) desc").Find(&ids)
```
[the remark this answers is not in the document]
Comment (lunny) [reply models/commit_status.go]: done.

File: routers/repo/setting_protected_branch.go (hunk 6 above)
Comment (lunny) [reply routers/repo/setting_protected_branch.go]: This has been removed.

File: routers/repo/setting_protected_branch.go (hunk 5 above)
Comment (lunny) [reply routers/repo/setting_protected_branch.go]: changed to 7*24*time.Hour

File: routers/repo/pull.go (hunk 4 above)
Comment (lunny) [reply routers/repo/pull.go]: done.

**Comment by lunny** [comment]
@techknowlogick @kolaente @guillep2k I think it's ready to review again.

**Review by guillep2k — APPROVED** [review]

**Comment by lunny** [comment]
@guillep2k done.

**Review by guillep2k — APPROVED** [review]

**Comment by lafriks** [comment]
@kolaente need your approval

**Comment by MarkusAmshove** [comment]
I want to discuss a scenario to see if it is supported or may introduce issues after release:

If a repository has multiple protected branches (think of version or customer branches), what are the options status checks?
If I look at the screenshots, a single commit status for ci/pr wouldn't be enough, because the check disambiguates the states by the name?
Would the ci need to send specific status like ci/pr/v1 ci/pr/v2 ?
If you think about Gitea, you might backport a commit to v1.9 and also target v1.10

**Comment by lunny** [comment]
@MarkusAmshove you can see the screenshot. All options here will be related to one branch(or branch rule in future). CI only needs to send a reponame(user+name) and a commit id. This will never be sent to v1.9 since it has been released.

**Comment by guillep2k** [comment]
> 
> 
> I want to discuss a scenario to see if it is supported or may introduce issues after release:
> 
> If a repository has multiple protected branches (think of version or customer branches), what are the options status checks?
> If I look at the screenshots, a single commit status for ci/pr wouldn't be enough, because the check disambiguates the states by the name?
> Would the ci need to send specific status like ci/pr/v1 ci/pr/v2 ?
> If you think about Gitea, you might backport a commit to v1.9 and also target v1.10

@MarkusAmshove If I understand correctly, you want to know if you can use Gitea when you have multiple CI scenarios for a repository (i.e. different branches). In Gitea, our CI is tied to our repository, and therefore to each push request. If we create a push request to another branch, a CI for that branch is used, although they share the same configuration most of the time.
If you want to implement different CI scripts depending on the branch you want to merge to, you'll have to use some kind of `if` in the webhook that launches your CI in order to use one or other configuration depending on the branch name. I don't think it would be very difficult.
I hope this helps.

**Comment by lunny** [comment]
@techknowlogick I updated this pull and added administrator bypass check, see screenshot 4 on issue content.

File: routers/repo/setting_protected_branch.go (hunk 7 above)
Comment (lafriks) [reply routers/repo/setting_protected_branch.go:188]: it still seems to left out

**Review by lafriks — CHANGES_REQUESTED** [review]
as per my comment

**Comment by MarkusAmshove** [comment]
@guillep2k I'm specifically asking because of the gitea plugin for Jenkins and if I have to do changes in there before I can use the status checks.

If I understand it correctly, the status which can be used for the PR checks are identified by the context name within the commit status table in the database.

So for each protected branch, I would want a different context for commit status.

For example, the gitea repository has version branches (1.8, 1.9, 1.10)

I create a PR from a bugfix branch and target 1.9.
The ci job for 1.9 will run and create a context `ci/drone/1.9` for the merge.
After it is merged, my branch also needs to be backported go 1.8, so I create another PR targeting 1.8.
The last PR will also show the status of 1.9, because it is associated with the commit, but it won't be a required check for 1.8.

Long story short, the context needs to be discriminated specifically for a target branch and not dynamically (e.g. drone/ci/pr124 and drone/ci/pr125)

File: routers/repo/setting_protected_branch.go (hunk 7 above)
Comment (lunny) [reply routers/repo/setting_protected_branch.go:188]: fixed again. :(

File: modules/pull/commit_status.go [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+	// check if all required status checks are successful
+	headGitRepo, err := git.OpenRepository(pr.HeadRepo.RepoPath())
+	if err != nil {
+		return false, errors.Wrap(err, "OpenRepository")
+	}
+
+	headBranchExist := headGitRepo.IsBranchExist(pr.HeadBranch)
+	if !headBranchExist {
+		return false, errors.New("HeadBranchExist is not exist, cannot merge")
```
Comment (lafriks) [inline modules/pull/commit_status.go]:
```suggestion
		return false, errors.New("Head branch does not exist, can not merge")
```

File: modules/pull/commit_status.go [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+	}
+
+	// check if all required status checks are successful
+	headGitRepo, err := git.OpenRepository(pr.HeadRepo.RepoPath())
+	if err != nil {
+		return false, errors.Wrap(err, "OpenRepository")
+	}
+
+	headBranchExist := headGitRepo.IsBranchExist(pr.HeadBranch)
+	if !headBranchExist {
```
Comment (lafriks) [inline modules/pull/commit_status.go]:
```suggestion
	if !headGitRepo.IsBranchExist(pr.HeadBranch) {
```

**Review by lafriks — CHANGES_REQUESTED** [review]

**Comment by lunny** [comment]
@lafriks all done.

**Review by lafriks — APPROVED** [review]
